staging/src/k8s.io/apiserver/pkg/server/options/authentication.go

```diff
@@ -235,59 +239,59 @@ const (
 	}
 
-	incluster, err := s.lookupInClusterClientCA()
+	client, err := s.getClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get delegated authentication kubeconfig: %v", err)
+	}
+
+	incluster, err := s.lookupInClusterClientCA(client)
 	if err != nil {
-		glog.Warningf("Unable to get configmap/%s in %s.  Usually fixed by "+
-			"'kubectl create rolebinding -n %s ROLE_NAME --role=%s --serviceaccount=YOUR_NS:YOUR_SA'",
-			authenticationConfigMapName, authenticationConfigMapNamespace, authenticationConfigMapNamespace, authenticationRoleName)
 		return nil, err
 	}
 	if incluster == nil {
-		return &s.ClientCert, ignorableError{fmt.Errorf("cluster doesn't provide client-ca-file in configmap/%s in %s, so client certificate authentication to extension api-server won't work.", authenticationConfigMapName, authenticationConfigMapNamespace)}
+		glog.Warningf("Cluster doesn't provide client-ca-file in configmap/%s in %s, so client certificate authentication to extension api-server won't work.", authenticationConfigMapName, authenticationConfigMapNamespace)
 	}
+
 	return incluster, nil
 }
 
-func (s *DelegatingAuthenticationOptions) getRequestHeader() (*RequestHeaderAuthenticationOptions, error) {
+func (s *DelegatingAuthenticationOptions) getRequestHeaderOrNil() (*RequestHeaderAuthenticationOptions, error) {
 	if len(s.RequestHeader.ClientCAFile) > 0 || s.SkipInClusterLookup {
 		return &s.RequestHeader, nil
 	}
 
-	incluster, err := s.lookupInClusterRequestHeader()
+	client, err := s.getClient()
 	if err != nil {
-		glog.Warningf("Unable to get configmap/%s in %s.  Usually fixed by "+
-			"'kubectl create rolebinding -n %s ROLE_NAME --role=%s --serviceaccount=YOUR_NS:YOUR_SA'",
-			authenticationConfigMapName, authenticationConfigMapNamespace, authenticationConfigMapNamespace, authenticationRoleName)
 		return nil, err
 	}
-	if incluster == nil {
-		return nil, fmt.Errorf("cluster doesn't provide requestheader-client-ca-file")
-	}
-	return incluster, nil
-}
 
-func (s *DelegatingAuthenticationOptions) lookupInClusterClientCA() (*ClientCertAuthenticationOptions, error) {
-	clientConfig, err := s.getClientConfig()
-	if err != nil {
-		return nil, fmt.Errorf("failed to get delegated authentication kubeconfig: %v", err)
-	}
-	client, err := coreclient.NewForConfig(clientConfig)
+	incluster, err := s.lookupInClusterRequestHeader(client)
 	if err != nil {
 		return nil, err
 	}
+	if incluster == nil {
+		glog.Warningf("Cluster doesn't provide requestheader-client-ca-file in configmap/%s in %s, so request-header client certificate authentication to extension api-server won't work.", authenticationConfigMapName, authenticationConfigMapNamespace)
+	}
+	return incluster, nil
+}
 
-	authConfigMap, err := client.ConfigMaps(authenticationConfigMapNamespace).Get(authenticationConfigMapName, metav1.GetOptions{})
+func (s *DelegatingAuthenticationOptions) lookupInClusterClientCA(client kubernetes.Interface) (*ClientCertAuthenticationOptions, error) {
```

Comment:
It's not visible in this PR. In a follow-up the client will be nil if no kubeconfig is passed and the component is not running in-cluster. The way it is factored now, the logic become much easier than handling a nil client inside the lookup funcs.

Reply:
Actually, we should pass the config map in here instead of the client. That removes duplication. Wdyt?

Reply:
> Actually, we should pass the config map in here instead of the client. That removes duplication. Wdyt?

good point. and these two will not return with errors.